docs/software-architecture-and-design.md

```diff
@@ -289,6 +294,57 @@ To summarize, the best practices are as follows:
+```
+src_state + event [guard] / action = dest_state
+```
+where the src\_state transitions to the dest\_state, while performing the _action_, only if the _event_ is processed and the _guard_ is true. Events are structs of new information that FSMs receive, so _guards_ and _actions_ take events as arguments. _Guards_ must return a boolean and _actions_ must return void. An asterix (\*) at the start of a row indicates that the state is an initial state.
+
+The library also supports hierarchical FSMs. Sub-FSMs are treated as states where an unconditional transition occurs when the sub-FSM is in the terminal state, X.
+```
+/* omitted rows of transition table */
+SubFSM = next_state, // Transitions to next_state only when the SubFSM is in the terminal state, X
+/* omitted rows of transition table */
+```
+In order to update a subFSM with an event, we need to do the following:
+```
+const auto update_sub_fsm_action =
+    [](auto event, back::process<TypeOfSubFSMEvent> processEvent) {
+        TypeOfSubFSMEvent sub_fsm_event = // initialize the subFSM event
+        processEvent(sub_fsm_event);
+    };
+```
+The convenience of this syntax comes at the cost of hard to read error messages due to the functor and templating system.
+
+## How Do We Use SML?
+We use SML to manage our [Tactics](#tactic). Each state represents a stage in the tactic where the robot should be doing a particular action or looking for certain conditions to be true. An example of this is the MoveFSM. While the robot is not at the destination and oriented correctly, the FSM is in the move state. Once the robot reaches its destination, it enters the terminal state, _X_, to indicate that it's done. SML also allows us to easily reuse FSMs in other tactics. For example, if a shadowing tactic needs to move to a particular destination with a certain orientation, then it can use the MoveFSM as a sub-FSM state.
+
+## SML Best Practices
+Boost-ext SML is a library that supports complex functionality with similarly complex syntax and semantics. If complex syntax is misused, the complicated error messages can make development difficult. Thus, we need to carefully choose a standardized subset of the library's syntax to implement our functionality while maintaining high readability.
+* Only use one _event_ per FSM: In gameplay, we react to changes in the [World](#world), so since there's only one source of new information, we should only need one _event_
+* Only one _guard_ or _action_ per transition: For readability of the transition table, we should only have one _guard_ or _action_ per transition. This can always be achieved by defining a special _guard_ or _action_ outside of the transition table.
```

Comment:
I'm not certain what makes a special guard "special" if it's defined out of the transition table. The way I read it first made it seem like the last sentence implied defining a special guard allowed you to have more than one guard which doesn't seem to be the case.

Reply:
good point